include/kframework/lemmas/bytes-simplification.k

```diff
@@ -3,6 +3,19 @@ module BYTES-SIMPLIFICATION [symbolic]
     imports EVM
     imports BUF
 
+  //
+  // concatHead/concatTail
+  //
+
+    syntax Bytes ::= concatHead(Bytes) [function, total] // Leftmost operand in a concat
+                   | concatTail(Bytes) [function, total] // Rest of concat
```

Comment:
Is this the right place to put this definition?

Reply:
Ah, I see it's to extract the left-most position now..... Hmmmmm, it seems to me that the same can be accomplished by just re-associating `_+Bytes_` with a rule like:

```
rule (BA1 +Bytes BA2) +Bytes BA3 => BA1 +Bytes (BA2 +Bytes BA3) [simplification]
```

Then have all the rules which use `concatHead` and `concatTail` just do direct pattern matching. Basically state in that we prefer right-associated byte-arrays.

Reply:
It is more general than pattern-matching, it is able to traverse arbitrarily associated concats. 

It seems to me as well that that re-association would isolate the head and the tail, but I am not sure what it would do to other simplifications (e.g., for combining ranges) that we use or how it would change the state in the contracts that we analysed.

I do welcome the idea of having a preferred associativity in principle.